plugin/flags_sources/makefile.py

```diff
@@ -0,0 +1,109 @@
+    """
+    _FILE_NAME = "Makefile"
+
+    def __init__(self, include_prefixes):
+        """Initialize a flag file storage.
+
+        Args:
+            include_prefixes (str[]): A List of valid include prefixes.
+        """
+        super().__init__(include_prefixes)
+        self._cache = MakefileCache()
+
+    def get_flags(self, file_path=None, search_scope=None):
+        """Get flags for file.
+
+        Args:
+            file_path (None, optional): A path to the query file.
+            search_scope (SearchScope, optional): Where to search for a
+                Makefile
+
+        Returns:
+            str[]: Return a list of flags in this Makefile
+        """
+        search_scope = self._update_search_scope(search_scope, file_path)
+        log.debug("[Makefile]:[get]: for file %s", file_path)
+        makefile_path = self._find_current_in(search_scope)
+        log.debug("[Makefile]:[current]: '%s'", makefile_path)
+        if not makefile_path:
+            return None
+        cached_makefile_path = self._get_cached_from(file_path)
+        log.debug("[Makefile]:[cached]: '%s'", cached_makefile_path)
+
+        if makefile_path in self._cache:
+            log.debug("[Makefile]: found cached Makefile")
+            cached_makefile_path = makefile_path
+            if makefile_path == cached_makefile_path:
+                if File.is_unchanged(cached_makefile_path):
+                    log.debug("[Makefile]:[unchanged]: load cached")
+                    return self._cache[cached_makefile_path]
+        log.debug("[Makefile]:[changed]: load new")
+        if cached_makefile_path and cached_makefile_path in self._cache:
+            del self._cache[cached_makefile_path]
+        flags = self.__flags_from_makefile(File(makefile_path))
+        if flags:
+            self._cache[makefile_path] = flags
+        if file_path:
+            self._cache[file_path] = makefile_path
+        return flags
+
+    def __flags_from_makefile(self, file):
+        """Get flags from Makefile.
+
+        Args:
+            file (File): A file objects that represents the file to parse.
+
+        Returns:
+            str[]: List of flags from Makefile.
+        """
+        if not path.exists(file.full_path()) or not file.loaded():
+            log.error("cannot get flags from Makefile. No file.")
+            return []
+
+        cmd = [
+            "make", "-s", "-C", file.folder(),
+            "-f", self._FILE_NAME, "-f", "-",
+        ]
+        makevars = [
+            "DEFAULT_INCLUDES",
+            "INCLUDES",
+            "AM_CPPFLAGS",
+            "CPPFLAGS",
+            "AM_CFLAGS",
+            "CFLAGS",
+        ]
+        for makevar in makevars:
+            cmd.append("print-" + makevar)
+        pipe = subprocess.PIPE
+        make = subprocess.Popen(cmd, stdout=pipe, stdin=pipe, stderr=pipe)
+        printer = "print-%:\n\t@echo '$($*)'\n".encode("utf-8")
+        output = make.communicate(input=printer)[0].decode("utf-8")
```

Comment:
Why not use the function from `Tools`?

https://github.com/niosus/EasyClangComplete/blob/34c3a9720d31b97f104e74eb945ce0775660486d/plugin/tools.py#L649-L650

Reply:
Mostly because of the use of `Popen.communicate()` to write to _make_'s stdin; `run_command()` is built upon `check_output()`, which AFAIK does not allow that.

So I guess we'd need to extend/change `run_command`, but directly using `Popen()` seems to be the simpler approach, imo.

Reply:
Makes sense. I am convinced.